File: src/com/kingmang/lino/parser/Parser.java
package com.kingmang.lino.parser;

import com.kingmang.lino.lexer.*;
import com.kingmang.lino.symbol.Environment;
import com.kingmang.lino.symbol.Symbol;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;

public class Parser {

    private StringBuffer arduinoBuffer;
    private Environment environment;
    private PrintWriter printWriter;
    private StringBuffer buffer;
    public StringBuilder console;
    private ArrayList<Integer> errors = new ArrayList<>();
    private Lexer lexer;
    private Token token;
    private String space = "";
    private boolean error;
    private int line;

    public Parser(Lexer lexer, File output) throws IOException {
        this.lexer = lexer;
        printWriter = new PrintWriter(output);
        buffer = new StringBuffer();
        arduinoBuffer = new StringBuffer();
        console = new StringBuilder();
        nextToken();
    }

    public String program() throws IOException {
        environment = null;
        arduinoBuffer.append(methods());

        printWriter.write(arduinoBuffer.toString());
        printWriter.close();

        appendLineToConsole(String.format("Done with %d %s", errors.size(), errors.size() <= 1 ? "error" : "errors"), errors.size() > 0);

        return arduinoBuffer.toString();

    }

    private String methods() throws IOException{
        StringBuilder methods = new StringBuilder();
        do {
            methods.append(method());
        } while (token.tag != TokenType.NULL);
        return methods.toString();
    }
    private String method() throws IOException {
        Environment saved = environment;
        environment = new Environment(environment);
        String tempString = space;
        space += '\t';
        boolean hasReturn;
        match(TokenType.FUNC);
        String name = match(TokenType.ID);
        match('(');
        String params = optParams();
        match(')');
        match(TokenType.COLON);
        StringBuilder returnType = new StringBuilder("void");
        if (token.tag == TokenType.TYPE) {
            returnType.delete(0,4);
            returnType.append(Generator.translateLexeme(match(TokenType.TYPE)));
            if (token.tag == '[') {
                returnType.append(match('['));
                returnType.append(match(']'));
            }
            hasReturn = true;
        } else if (token.tag == TokenType.VOID) {
            match(TokenType.VOID);
            hasReturn = false;
        } else {
            error(TokenType.RETURN, "return");
            hasReturn = true;
        }
        match(TokenType.LBRACE);
        String decelerations = decelerations();
        String statements = optStatement();

        StringBuilder optReturn = new StringBuilder("\t");
        if (hasReturn) {
            optReturn.append(space);
            optReturn.append("\n\t").append(match(TokenType.RETURN)).append(" ");
            optReturn.append(expr());
            optReturn.append(';');
            optReturn.append("\n\t");
        } else if (token.tag == TokenType.RETURN) {
            optReturn.append(space);
            optReturn.append(match(TokenType.RETURN));
            optReturn.append(';');
            optReturn.append("\n\t");
        }
        match(TokenType.RBRACE);

        String template = "%s%s(%s){\n%s%s%s\n}\n";
        environment = saved;
        space = tempString;
        return String.format(template, returnType, name, params, decelerations, statements, optReturn);
    }

    private String optStatement() throws IOException{
        StringBuilder optStatement = new StringBuilder();
        while(token.tag != TokenType.RETURN && token.tag != TokenType.RBRACE && token.tag != TokenType.NULL){
            optStatement.append(statement());
        }
        return optStatement.toString();

    }
    private String optParams() throws IOException {
        StringBuilder optParams = new StringBuilder();
        optParams.append(param());
        while (true) {
            if (token.tag == ',') {
                optParams.append(match(','));
                optParams.append(param());
            } else return optParams.toString();
        }
    }
    private String decelerations() throws IOException {
        StringBuilder decelerations = new StringBuilder();
        while (true) {
            if (token.tag == TokenType.TYPE) decelerations.append(deceleration());
            else break;
        }
        return decelerations.toString();
    }
    private String deceleration() throws IOException {
        StringBuilder decelration = new StringBuilder();
        Symbol symbol = new Symbol();
        String type = match(TokenType.TYPE);
        symbol.setType(type);
        decelration.append(Generator.translateLexeme(type));
        String id = match(TokenType.ID);
        decelration.append(id);
        if (token.tag == '[') {
            decelration.append(match('['));
            decelration.append(match(']'));
        }
        if (token.tag == TokenType.ASSIGN) {
            decelration.append(match(TokenType.ASSIGN));
            String value = expr();
            decelration.append(value);
            symbol.setValue(value);
        }
        if (environment.get(id) == null)
            environment.put(id, symbol);
        else appendLineToConsole("Error: Variable ".concat(id).concat(" is already defined in the scope"), true);
        decelration.append(';');
        return space + decelration.toString() + "\n";
    }
    private String param() throws IOException{
        StringBuilder param = new StringBuilder();
        if(token.tag == TokenType.TYPE) {
            Symbol symbol = new Symbol();
            String type = match(TokenType.TYPE);
            symbol.setType(type);
            param.append(Generator.translateLexeme(type));
            String id = match(TokenType.ID);
            param.append(id);
            if (token.tag == '[') {
                param.append(match('['));
                param.append(match(']'));
            }
            if (environment.get(id) == null) {
                environment.put(id, symbol);
            }else {
                appendLineToConsole(
                        "Error: Variable"
                                .concat(id)
                                .concat("is already definited in the scope"),
                        true);
            }
        }else if(token.tag == TokenType.LITERAl){
            param.append(match(TokenType.LITERAl));
        }
        return param.toString();
    }

    private String callParams() throws IOException{
        StringBuilder callParams = new StringBuilder();
        if(token.tag == '(' || token.tag == TokenType.NUM || token instanceof Word || token.tag == TokenType.LITERAl){
            callParams.append(expr());
            while (true){
                if(token.tag == ','){
                    callParams.append(match(','));
                    callParams.append(expr());
                }else break;
            }
        }
        return callParams.toString();
    }
    private String statement() throws IOException {
        String tempString = space;
        StringBuilder statement = new StringBuilder();
        if (token.tag == TokenType.IF) {
            statement.append(match(TokenType.IF));
            space += '\t';
            statement.append('(');
            statement.append(booleanExpr());
            match(TokenType.LBRACE);
            statement.append(')').append(" {\n ");
            statement.append(optStatement());
            match(TokenType.RBRACE);
            space = tempString;
            statement.append(space).append('}');
        } else if (token.tag == TokenType.WHILE) {
            statement.append(match(TokenType.WHILE));
            space += '\t';
            statement.append('(');
            statement.append(booleanExpr());
            statement.append(')').append(" {\n ");
            match(TokenType.LBRACE);
            statement.append(optStatement());
            match(TokenType.RBRACE);
            space = tempString;
            statement.append(space).append("}");
        } else if (token.tag == TokenType.REPEAT) {
            match(TokenType.REPEAT);
            space += '\t';
            statement.append("for (int i = 0; i < ");
            statement.append(expr());
            statement.append("; i++) {\n");
            match(TokenType.LBRACE);
            statement.append(optStatement());
            match(TokenType.RBRACE);
            space = tempString;
            statement.append(space).append("}");
        } else if (token.tag == TokenType.ID) {
            String id = match(TokenType.ID);
            String value;
            statement.append(id);
            if (token.tag == '(') {
                statement.append(match('('));
                statement.append(callParams());
                statement.append(match(')'));
                statement.append(match(';'));
            } else {
                if (token.tag == '[') {
                    statement.append(match('['));
                    statement.append(expr());
                    statement.append(match(']'));
                }
                statement.append(match(TokenType.ASSIGN));
                value = expr();
                statement.append(value);
                Symbol symbol = environment.get(id);
                if (symbol != null) {
                    symbol.setValue(value);
                } else {
                    appendLineToConsole(String.format("Error at line: %d Cannot resolve symbol %s", lexer.line, id), true);
                }
                statement.append(';');
            }
        } else if (token.tag == TokenType.PRINT) {
            match(TokenType.PRINT);
            statement.append("Serial.print");
            statement.append('(');
            statement.append(expr());
            statement.append(')');
            statement.append(';');
        } else {
            appendLineToConsole("line: ".concat(String.valueOf(lexer.line)).concat(" Not A Statement"), true);
            nextToken();
        }
        error = false;

        return space + statement.toString();
    }

    private String booleanExpr() throws IOException {
        StringBuilder booleanExpr = new StringBuilder();
        booleanExpr.append(expr());
        if (token.tag == TokenType.COMPARATOR) {
            booleanExpr.append(Generator.translateLexeme(match(TokenType.COMPARATOR)));
            booleanExpr.append(expr());
        }
        return booleanExpr.toString();
    }

    private String expr() throws IOException {
        StringBuilder expr = new StringBuilder();
        expr.append(term());
        while (true) {
            if (token.tag == '+') {
                expr.append(match('+'));
                expr.append(term());
                if (!error) {
                    buffer.append("+ ");
                }
            } else if (token.tag == '-') {
                expr.append(match('-'));
                expr.append(term());
                if (!error) {
                    buffer.append("- ");
                }
            } else return expr.toString();
        }
    }

    private String term() throws IOException {
        StringBuilder term = new StringBuilder();
        term.append(factor());
        while (true) {
            if (token.tag == '*') {
                term.append(match('*'));
                term.append(factor());
                if (!error) {
                    buffer.append("* ");
                }
            } else if (token.tag == '/') {
                term.append(match('/'));
                term.append(factor());
                if (!error) {
                    buffer.append("/ ");
                }
            } else return term.toString();
        }
    }

    private String factor() throws IOException {
        StringBuilder factor = new StringBuilder();
        if (token.tag == '(') {
            factor.append(match('('));
            factor.append(expr());
            factor.append(match(')'));
        } else if (token instanceof Num) {
            buffer.append(((Num) token).value).append(' ');
            factor.append(match(TokenType.NUM));
        } else if (token.tag == TokenType.ID) {
            buffer.append(((Word) token).getLexeme()).append(' ');
            String id = match(TokenType.ID);
            factor.append(id);
            if (token.tag == '.') {
                factor.append(match('.'));
                factor.append(match(TokenType.ID));
            }
            if (token.tag == '[') {
                factor.append(match('['));
                factor.append(expr());
                factor.append(match(']'));
            }
            if (environment.get(id) == null) {
                appendLineToConsole(String.format("Error at line: %d Cannot resolve symbol %s", lexer.line, id), true);
            }
        } else if (token.tag == TokenType.LITERAl) {
            factor.append(match(TokenType.LITERAl));
        } else {
            error(TokenType.NUM, "NUM/Word/LITERAL");
        }
        return factor.toString();
    }

    private String match(int matchTag) throws IOException{
        if(token.tag == matchTag){
            Token temp = token;
            nextToken();
            return temp.stringify();
        }else{
            error(matchTag, String.valueOf(matchTag));
            return "";
        }
    }
    private void error(int tag, String expected) throws IOException{
        if(!error){
            error = true;
            int last = buffer.lastIndexOf("\n");
            if(last >= 0){
                buffer.delete(last, buffer.length());
                appendLineToConsole("Expected".concat("Error at line: ".concat(String.valueOf(lexer.line)).concat("").concat(expected)), true);
                skipErrors(new Token(tag));
            }
        }
    }
    private void skipErrors(Token expected) throws IOException {
        if(!isEndToken(expected) && !isEndToken(token)){
            do{
                nextToken();
            }while(!isEndToken(token));
        }
    }
    private void nextToken() throws IOException {
        do{
            token = lexer.tokenize();
        }while (token.tag == TokenType.COMMENT);
        appendLineToConsole(token.toString(), false);
    }
    private boolean isEndToken(Token token){
        return
                token.tag == ';' ||
                token.tag == TokenType.NULL ||
                token.tag == TokenType.RBRACE ||
                token.tag == TokenType.RETURN;
    }
    private void appendLineToConsole(String string, boolean isError){
        console.append(string).append("\n");
        if(isError){
            errors.add(this.line);
        }
        this.line++;
    }
}

File: parser/parser.go
package parser

import (
	"bytes"
	"fmt"
	"strconv"
	"strings"

	"github.com/kingmang/lino/lexer"
	"github.com/kingmang/lino/symbol"
)

type Parser struct {
	lexer      *lexer.Lexer
	token      lexer.Token
	output     strings.Builder
	env        *symbol.Environment
	console    strings.Builder
	errors     []int
	space      string
	hasError   bool
	line       int
	bufBuilder bytes.Buffer
}

func NewParser(l *lexer.Lexer) *Parser {
	p := &Parser{lexer: l}
	p.nextToken()
	return p
}

func (p *Parser) Program() error {
	p.env = nil
	p.output.WriteString(p.methods())

	errCount := len(p.errors)
	errWord := "errors"
	if errCount <= 1 {
		errWord = "error"
	}
	p.appendLineToConsole(fmt.Sprintf("Done with %d %s", errCount, errWord), errCount > 0)

	return nil
}

func (p *Parser) Output() string {
	return p.output.String()
}

func (p *Parser) Console() string {
	return p.console.String()
}

func (p *Parser) methods() string {
	var methods strings.Builder
	for {
		methods.WriteString(p.method())
		if p.token.GetTag() == lexer.EOF {
			break
		}
	}
	return methods.String()
}

func (p *Parser) method() string {
	saved := p.env
	p.env = symbol.NewEnvironment(p.env)
	tempSpace := p.space
	p.space += "\t"
	hasReturn := false

	p.match(lexer.FUNC)
	name := p.match(lexer.ID)
	p.match('(')
	params := p.optParams()
	p.match(')')
	p.match(lexer.COLON)

	returnType := "void"
	if p.token.GetTag() == lexer.TYPE {
		returnType = TranslateLexeme(p.match(lexer.TYPE))
		if p.token.GetTag() == '[' {
			returnType += p.match('[')
			returnType += p.match(']')
		}
		hasReturn = true
	} else if p.token.GetTag() == lexer.VOID {
		p.match(lexer.VOID)
		hasReturn = false
	} else {
		p.error(lexer.RETURN, "return")
		hasReturn = true
	}

	p.match(lexer.LBRACE)
	decelerations := p.decelerations()
	statements := p.optStatement()

	var optReturn strings.Builder
	optReturn.WriteString("\t")
	if hasReturn {
		optReturn.WriteString(p.space)
		optReturn.WriteString("\n\t")
		optReturn.WriteString(p.match(lexer.RETURN))
		optReturn.WriteString(" ")
		optReturn.WriteString(p.expr())
		optReturn.WriteString(";")
		optReturn.WriteString("\n\t")
	} else if p.token.GetTag() == lexer.RETURN {
		optReturn.WriteString(p.space)
		optReturn.WriteString(p.match(lexer.RETURN))
		optReturn.WriteString(";")
		optReturn.WriteString("\n\t")
	}
	p.match(lexer.RBRACE)

	p.env = saved
	p.space = tempSpace

	return fmt.Sprintf("%s%s(%s){\n%s%s%s\n}\n", returnType, name, params, decelerations, statements, optReturn.String())
}

func (p *Parser) optStatement() string {
	var optStatement strings.Builder
	for p.token.GetTag() != lexer.RETURN && p.token.GetTag() != lexer.RBRACE && p.token.GetTag() != lexer.EOF {
		optStatement.WriteString(p.statement())
	}
	return optStatement.String()
}

func (p *Parser) optParams() string {
	var optParams strings.Builder
	optParams.WriteString(p.param())
	for {
		if p.token.GetTag() == ',' {
			optParams.WriteString(p.match(','))
			optParams.WriteString(p.param())
		} else {
			return optParams.String()
		}
	}
}

func (p *Parser) decelerations() string {
	var decelerations strings.Builder
	for {
		if p.token.GetTag() == lexer.TYPE {
			decelerations.WriteString(p.deceleration())
		} else {
			break
		}
	}
	return decelerations.String()
}

func (p *Parser) deceleration() string {
	var decelration strings.Builder
	sym := &symbol.Symbol{}
	typeName := p.match(lexer.TYPE)
	sym.Type = typeName
	decelration.WriteString(TranslateLexeme(typeName))
	id := p.match(lexer.ID)
	decelration.WriteString(id)
	if p.token.GetTag() == '[' {
		decelration.WriteString(p.match('['))
		decelration.WriteString(p.match(']'))
	}
	if p.token.GetTag() == lexer.ASSIGN {
		decelration.WriteString(p.match(lexer.ASSIGN))
		value := p.expr()
		decelration.WriteString(value)
		sym.Value = value
	}
	if p.env.Get(id) == nil {
		p.env.Put(id, sym)
	} else {
		p.appendLineToConsole("Error: Variable "+id+" is already defined in the scope", true)
	}
	decelration.WriteString(";")
	return p.space + decelration.String() + "\n"
}

func (p *Parser) param() string {
	var param strings.Builder
	if p.token.GetTag() == lexer.TYPE {
		sym := &symbol.Symbol{}
		typeName := p.match(lexer.TYPE)
		sym.Type = typeName
		param.WriteString(TranslateLexeme(typeName))
		id := p.match(lexer.ID)
		param.WriteString(id)
		if p.token.GetTag() == '[' {
			param.WriteString(p.match('['))
			param.WriteString(p.match(']'))
		}
		if p.env.Get(id) == nil {
			p.env.Put(id, sym)
		} else {
			p.appendLineToConsole("Error: Variable"+id+"is already definited in the scope", true)
		}
	} else if p.token.GetTag() == lexer.LITERAL {
		param.WriteString(p.match(lexer.LITERAL))
	}
	return param.String()
}

func (p *Parser) callParams() string {
	var callParams strings.Builder
	_, isWord := p.token.(*lexer.WordToken)
	if p.token.GetTag() == '(' || p.token.GetTag() == lexer.NUM || isWord || p.token.GetTag() == lexer.LITERAL {
		callParams.WriteString(p.expr())
		for {
			if p.token.GetTag() == ',' {
				callParams.WriteString(p.match(','))
				callParams.WriteString(p.expr())
			} else {
				break
			}
		}
	}
	return callParams.String()
}

func (p *Parser) statement() string {
	tempSpace := p.space
	var statement strings.Builder

	if p.token.GetTag() == lexer.IF {
		statement.WriteString(p.match(lexer.IF))
		p.space += "\t"
		statement.WriteString("(")
		statement.WriteString(p.booleanExpr())
		p.match(lexer.LBRACE)
		statement.WriteString(") {\n ")
		statement.WriteString(p.optStatement())
		p.match(lexer.RBRACE)
		p.space = tempSpace
		statement.WriteString(p.space)
		statement.WriteString("}")
	} else if p.token.GetTag() == lexer.WHILE {
		statement.WriteString(p.match(lexer.WHILE))
		p.space += "\t"
		statement.WriteString("(")
		statement.WriteString(p.booleanExpr())
		statement.WriteString(") {\n ")
		p.match(lexer.LBRACE)
		statement.WriteString(p.optStatement())
		p.match(lexer.RBRACE)
		p.space = tempSpace
		statement.WriteString(p.space)
		statement.WriteString("}")
	} else if p.token.GetTag() == lexer.REPEAT {
		p.match(lexer.REPEAT)
		p.space += "\t"
		statement.WriteString("for (int i = 0; i < ")
		statement.WriteString(p.expr())
		statement.WriteString("; i++) {\n")
		p.match(lexer.LBRACE)
		statement.WriteString(p.optStatement())
		p.match(lexer.RBRACE)
		p.space = tempSpace
		statement.WriteString(p.space)
		statement.WriteString("}")
	} else if p.token.GetTag() == lexer.ID {
		id := p.match(lexer.ID)
		statement.WriteString(id)
		if p.token.GetTag() == '(' {
			statement.WriteString(p.match('('))
			statement.WriteString(p.callParams())
			statement.WriteString(p.match(')'))
			statement.WriteString(p.match(';'))
		} else {
			if p.token.GetTag() == '[' {
				statement.WriteString(p.match('['))
				statement.WriteString(p.expr())
				statement.WriteString(p.match(']'))
			}
			statement.WriteString(p.match(lexer.ASSIGN))
			value := p.expr()
			statement.WriteString(value)
			sym := p.env.Get(id)
			if sym != nil {
				sym.Value = value
			} else {
				p.appendLineToConsole(fmt.Sprintf("Error at line: %d Cannot resolve symbol %s", p.lexer.Line, id), true)
			}
			statement.WriteString(";")
		}
	} else if p.token.GetTag() == lexer.PRINT {
		p.match(lexer.PRINT)
		statement.WriteString("Serial.print")
		statement.WriteString("(")
		statement.WriteString(p.expr())
		statement.WriteString(")")
		statement.WriteString(";")
	} else {
		p.appendLineToConsole("line: "+strconv.Itoa(p.lexer.Line)+" Not A Statement", true)
		p.nextToken()
	}
	p.hasError = false

	return p.space + statement.String()
}

func (p *Parser) booleanExpr() string {
	var booleanExpr strings.Builder
	booleanExpr.WriteString(p.expr())
	if p.token.GetTag() == lexer.COMPARATOR {
		booleanExpr.WriteString(TranslateLexeme(p.match(lexer.COMPARATOR)))
		booleanExpr.WriteString(p.expr())
	}
	return booleanExpr.String()
}

func (p *Parser) expr() string {
	var expr strings.Builder
	expr.WriteString(p.term())
	for {
		if p.token.GetTag() == '+' {
			expr.WriteString(p.match('+'))
			expr.WriteString(p.term())
			if !p.hasError {
				p.bufBuilder.WriteString("+ ")
			}
		} else if p.token.GetTag() == '-' {
			expr.WriteString(p.match('-'))
			expr.WriteString(p.term())
			if !p.hasError {
				p.bufBuilder.WriteString("- ")
			}
		} else {
			return expr.String()
		}
	}
}

func (p *Parser) term() string {
	var term strings.Builder
	term.WriteString(p.factor())
	for {
		if p.token.GetTag() == '*' {
			term.WriteString(p.match('*'))
			term.WriteString(p.factor())
			if !p.hasError {
				p.bufBuilder.WriteString("* ")
			}
		} else if p.token.GetTag() == '/' {
			term.WriteString(p.match('/'))
			term.WriteString(p.factor())
			if !p.hasError {
				p.bufBuilder.WriteString("/ ")
			}
		} else {
			return term.String()
		}
	}
}

func (p *Parser) factor() string {
	var factor strings.Builder
	if p.token.GetTag() == '(' {
		factor.WriteString(p.match('('))
		factor.WriteString(p.expr())
		factor.WriteString(p.match(')'))
	} else if num, ok := p.token.(*lexer.NumToken); ok {
		p.bufBuilder.WriteString(strconv.Itoa(num.Value))
		p.bufBuilder.WriteString(" ")
		factor.WriteString(p.match(lexer.NUM))
	} else if p.token.GetTag() == lexer.ID {
		word := p.token.(*lexer.WordToken)
		p.bufBuilder.WriteString(word.Lexeme)
		p.bufBuilder.WriteString(" ")
		id := p.match(lexer.ID)
		factor.WriteString(id)
		if p.token.GetTag() == '.' {
			factor.WriteString(p.match('.'))
			factor.WriteString(p.match(lexer.ID))
		}
		if p.token.GetTag() == '[' {
			factor.WriteString(p.match('['))
			factor.WriteString(p.expr())
			factor.WriteString(p.match(']'))
		}
		if p.env.Get(id) == nil {
			p.appendLineToConsole(fmt.Sprintf("Error at line: %d Cannot resolve symbol %s", p.lexer.Line, id), true)
		}
	} else if p.token.GetTag() == lexer.LITERAL {
		factor.WriteString(p.match(lexer.LITERAL))
	} else {
		p.error(lexer.NUM, "NUM/Word/LITERAL")
	}
	return factor.String()
}

func (p *Parser) match(matchTag int) string {
	if p.token.GetTag() == matchTag {
		temp := p.token
		p.nextToken()
		return temp.Stringify()
	} else {
		p.error(matchTag, strconv.Itoa(matchTag))
		return ""
	}
}

// tag not used
func (p *Parser) error(_ int, expected string) {
	if !p.hasError {
		p.hasError = true
		last := bytes.LastIndex(p.bufBuilder.Bytes(), []byte("\n"))
		if last >= 0 {
			p.bufBuilder.Truncate(last)
			p.appendLineToConsole("ExpectedError at line: "+strconv.Itoa(p.lexer.Line)+""+expected, true)
			p.skipErrors(&lexer.BaseToken{})
		}
	}
}

func (p *Parser) skipErrors(expected lexer.Token) {
	if p.isEndToken(expected) && p.isEndToken(p.token) {
		for {
			p.nextToken()
			if !p.isEndToken(p.token) {
				break
			}
		}
	}
}

func (p *Parser) nextToken() {
	for {
		t, err := p.lexer.Tokenize()
		if err != nil {
			p.token = &lexer.BaseToken{Tag: lexer.EOF}
			return
		}
		p.token = t
		if p.token.GetTag() != lexer.COMMENT {
			break
		}
	}
	p.appendLineToConsole(p.token.String(), false)
}

func (p *Parser) isEndToken(token lexer.Token) bool {
	return token.GetTag() != ';' &&
		token.GetTag() != lexer.EOF &&
		token.GetTag() != lexer.RBRACE &&
		token.GetTag() != lexer.RETURN
}

func (p *Parser) appendLineToConsole(str string, isError bool) {
	p.console.WriteString(str)
	p.console.WriteString("\n")
	if isError {
		p.errors = append(p.errors, p.line)
	}
	p.line++
}
